crates/oxc_linter/src/rules/oxc/no_this_in_exported_function.rs

```diff
@@ -0,0 +1,189 @@
+use oxc_ast::{
+    AstKind,
+    ast::{
+        Declaration, ExportDefaultDeclarationKind, Expression, Function, ModuleExportName,
+        ThisExpression,
+    },
+};
+use oxc_ast_visit::Visit;
+use oxc_diagnostics::OxcDiagnostic;
+use oxc_macros::declare_oxc_lint;
+use oxc_semantic::ScopeFlags;
+use oxc_span::Span;
+
+use crate::{AstNode, context::LintContext, rule::Rule};
+
+fn no_this_in_exported_function_diagnostic(span: Span) -> OxcDiagnostic {
+    OxcDiagnostic::warn("`this` should not be used in exported functions")
+        .with_help("Remove `this` or convert to a non-exported function. In bundlers, `this` becomes `undefined` in exported functions.")
+        .with_label(span)
+}
+
+#[derive(Debug, Default, Clone)]
+pub struct NoThisInExportedFunction;
+
+declare_oxc_lint!(
+    /// ### What it does
+    ///
+    /// Disallows the use of `this` in exported functions.
+    ///
+    /// ### Why is this bad?
+    ///
+    /// In most bundlers, the value of `this` is not preserved for exported functions.
+    /// When a function is exported and imported in another module, `this` typically
+    /// becomes `undefined` instead of the module namespace object. This can lead to
+    /// unexpected runtime errors or incorrect behavior.
+    ///
+    /// ### Examples
+    ///
+    /// Examples of **incorrect** code for this rule:
+    /// ```javascript
+    /// export function foo() {
```

Comment:
@sapphi-red 

I wonder whether we should explicitly allow usage of `this` if it's specified as a parameter

`export function foo(this: SomeType)) {`

(it makes the assumption that the function is bound where ever it's used)

We don't need to do it in this PR though